internal/channel/plugin.go

```diff
@@ -27,60 +83,96 @@ func (c *Channel) GetInfo() (*plugin.Info, error) {
 	return info, nil
 }
 
-func (c *Channel) SetConfig(config string) error {
-	_, err := c.rpcCall(plugin.MethodSetConfig, json.RawMessage(config))
+// SetConfig sends the setConfig request with given config, returns an error if an error occurred
+func (p *Plugin) SetConfig(config string) error {
+	_, err := p.rpc.Call(plugin.MethodSetConfig, json.RawMessage(config))
 
 	return err
 }
 
-func (c *Channel) SendNotification(contact *recipient.Contact, i contracts.Incident, ev *event.Event, icingaweb2Url string) error {
-	contactStruct := &plugin.Contact{FullName: contact.FullName}
-	for _, addr := range contact.Addresses {
-		contactStruct.Addresses = append(contactStruct.Addresses, &plugin.Address{Type: addr.Type, Address: addr.Address})
+// SendNotification sends the notification, returns an error if fails
+func (p *Plugin) SendNotification(req *plugin.NotificationRequest) error {
+	params, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("failed to prepare request params: %w", err)
 	}
 
-	if !strings.HasSuffix(icingaweb2Url, "/") {
-		icingaweb2Url += "/"
-	}
+	_, err = p.rpc.Call(plugin.MethodSendNotification, params)
+
+	return err
+}
+
+func terminate(p *Plugin) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.logger.Debug("Stopping the channel plugin")
+
+	_ = p.rpc.Close()
+	timer := time.AfterFunc(5*time.Second, func() {
+		p.logger.Debug("killing the channel plugin")
+		_ = p.cmd.Process.Kill()
+	})
+
+	<-p.rpc.Done()
+	timer.Stop()
+	p.logger.Warn("Stopped channel plugin")
+}
 
-	req := plugin.NotificationRequest{
-		Contact: contactStruct,
-		Object: &plugin.Object{
-			Name:      i.ObjectDisplayName(),
-			Url:       ev.URL,
-			Tags:      ev.Tags,
-			ExtraTags: ev.ExtraTags,
-		},
-		Incident: &plugin.Incident{
-			Id:  i.ID(),
-			Url: fmt.Sprintf("%snotifications/incident?id=%d", icingaweb2Url, i.ID()),
-		},
-		Event: &plugin.Event{
-			Time:     ev.Time,
-			Type:     ev.Type,
-			Severity: ev.Severity.String(),
-			Username: ev.Username,
-			Message:  ev.Message,
-		},
+func forwardLogs(errPipe io.Reader, logger *zap.SugaredLogger) {
+	reader := bufio.NewReader(errPipe)
+	for {
+		line, err := reader.ReadString('\n')
+		if err != nil {
+			if err != io.EOF {
+				logger.Errorw("Failed to read stderr line", zap.Error(err))
+			}
+
+			return
+		}
+
+		logger.Info(line)
 	}
+}
 
-	params, err := json.Marshal(req)
+// SyncPlugins upsert the plugin table with working plugins
+func SyncPlugins(channelPluginDir string, logger *logging.Logger, db *icingadb.DB) {
+	files, err := os.ReadDir(channelPluginDir)
```

Comment:
Yes and no. I agree that this can be made more robust to avoid unintended behavior (exclude hidden/backup files (`.*`, `*~`) or maybe require more specific file names, check permissions, ...), but is the overall idea worse than pretty much any `*.d` config directory?

I mean we could consider something like store what's returned from `GetInfo()` currently in a separate file, so that for example you'd have the `email` binary and a `email.info` file simply containing JSON, that would save us from executing the plugins until they are actually configured, but it wouldn't change much about "if it's in this directory, it's trusted".

Reply:
Should i leave it as is for now?

Reply:
Fair enough.

However, I would still check within the `for` loop over those `files` if each file is really a file. Otherwise, when accidentally creating a subdirectory within the hopefully trusted plugin directory, creating the `NewPlugin` will fail. Furthermore, I would also deny symlinks (_or anything not being a file_) as those would dangerously widen the trusted plugin source.

Reply:
As discussed offline, calling `NewPlugin` will fail in all the above cases, so let's leave it as is for now.